core/pbcc_subscribe_v2.c

```diff
@@ -0,0 +1,259 @@
+#include "pubnub_internal.h"
+
+#include "pubnub_version.h"
+#include "pubnub_assert.h"
+#include "pubnub_json_parse.h"
+#include "pubnub_log.h"
+#include "pubnub_url_encode.h"
+#include "lib/pb_strnlen_s.h"
+#include "pbcc_subscribe_v2.h"
+
+
+#include <stdio.h>
+#include <stdlib.h>
+
+
+/** It has to contain the `t` field, with another `t` for the
+    timetoken (as string) and `tr` for the region (integer) and the
+    `m` field for the message(s) array.
+*/
+#define MIN_SUBSCRIBE_V2_RESPONSE_LENGTH 40
+
+enum pubnub_res pbcc_subscribe_v2_prep(struct pbcc_context* p,
+                                       char const*          channel,
+                                       char const*          channel_group,
+                                       unsigned*            heartbeat,
+                                       char const*          filter_expr)
+{
+    char        region_str[20];
+    char const* tr;
+
+    if (NULL == channel) {
+        if (NULL == channel_group) {
+            return PNR_INVALID_CHANNEL;
+        }
+        channel = ",";
+    }
+    if (p->msg_ofs < p->msg_end) {
+        return PNR_RX_BUFF_NOT_EMPTY;
+    }
+
+    if ('\0' == p->timetoken[0]) {
+        p->timetoken[0] = '0';
+        p->timetoken[1] = '\0';
+        tr              = NULL;
+    }
+    else {
+        snprintf(region_str, sizeof region_str, "%d", p->region);
+        tr = region_str;
+    }
+    p->http_content_len = 0;
+    p->msg_ofs = p->msg_end = 0;
+
+    p->http_buf_len = snprintf(
+        p->http_buf, sizeof p->http_buf, "/v2/subscribe/%s/", p->subscribe_key);
+    APPEND_URL_ENCODED_M(p, channel);
+    p->http_buf_len += snprintf(p->http_buf + p->http_buf_len,
+                                sizeof p->http_buf - p->http_buf_len,
+                                "/0?tt=%s&pnsdk=%s",
+                                p->timetoken,
+                                pubnub_uname());
+    APPEND_URL_PARAM_M(p, "tr", tr, '&');
+    APPEND_URL_PARAM_M(p, "channel-group", channel_group, '&');
+    APPEND_URL_PARAM_M(p, "uuid", p->uuid, '&');
+    APPEND_URL_PARAM_M(p, "auth", p->auth, '&');
+    APPEND_URL_PARAM_ENCODED_M(p, "filter-expr", filter_expr, '&');
+    APPEND_URL_OPT_PARAM_UNSIGNED_M(p, "heartbeat", heartbeat, '&');
+
+    return PNR_STARTED;
+}
+
+
+enum pubnub_res pbcc_parse_subscribe_v2_response(struct pbcc_context* p)
+{
+    enum pbjson_object_name_parse_result jpresult;
+    struct pbjson_elem                   el;
+    struct pbjson_elem                   found;
+    char*                                reply = p->http_reply;
+
+    if (p->http_buf_len < MIN_SUBSCRIBE_V2_RESPONSE_LENGTH) {
+        return PNR_FORMAT_ERROR;
+    }
+    if ((reply[0] != '{') || (reply[p->http_buf_len - 1] != '}')) {
+        return PNR_FORMAT_ERROR;
+    }
+
+    el.start = p->http_reply;
+    el.end   = p->http_reply + p->http_buf_len;
+    jpresult = pbjson_get_object_value(&el, "t", &found);
+    if (jonmpOK == jpresult) {
```

Comment:
Is this a typo in the variable `jonmpOK` ? Should it actually be `jsonpOK` ? Or something like that.

Reply:
`jonmp` here is a common prefix for the enum that has results for our "JsON Mini Parser".

Reply:
lol thanks for the explanation